File: DotNet/Conversations/Conversation.cs
using System;
using System.Collections.Generic;
using System.Reactive;
using System.Reactive.Subjects;
using Shastra.Hydra.Messaging;
using Shastra.Hydra.Messaging.Attachments;
using Shastra.Hydra.Messaging.MessageIds;
using Shastra.Hydra.Messaging.Serializers;

namespace Shastra.Hydra.Conversations
{
    public class Conversation<TMessage> : IObservable<Notification<AugmentedMessage<TMessage>>>, IDisposable
    {
        private ISerializer<TMessage> _serializer;
        private readonly Subject<Notification<AugmentedMessage<TMessage>>> _subject = new Subject<Notification<AugmentedMessage<TMessage>>>();
        private bool _done;
        private IHydraService _hydraService;

        public string Handle { get; private set; }
        public string RemoteParty { get; private set; }
        public string ThisParty { get; private set; }
        public string Topic { get; private set; }
        public long LastSendSeq { get; private set; }
        public long LastRecvSeq { get; private set; }
        public bool CheckSeq { get; set; }

        internal event Action<object> DoneEvent;

        internal void BaseInit(IHydraService hydraService, string thisParty, string remoteParty, string topic, string handle, ISerializer<TMessage> serializer)
        {
            _hydraService = hydraService;
            ThisParty = thisParty;
            RemoteParty = remoteParty;
            Topic = topic;
            Handle = handle;
            _serializer = serializer;
        }

        internal void OnNext(long seq, Notification<AugmentedMessage<TMessage>> message)
        {
            if (CheckSeq && seq != LastRecvSeq + 1) {
                // Tell the client, but carry on and process the message.
                _subject.OnError(new Exception(string.Format("Sequence error. Expected {0}, received {1}", LastRecvSeq + 1, seq)));
            }
            LastRecvSeq = seq;
            _subject.OnNext(message);
        }

        /// <summary>
        /// Send a message to the other end of the conversation, with optional attachments
        /// </summary>
        /// <param name="message">The message to send</param>
        /// <param name="attachments">Attachments to send with the message</param>
        /// <returns>The id of the message sent</returns>
        public IMessageId Send(TMessage message, IEnumerable<Attachment> attachments = null)
        {
            if (_done) return null;

            var hydraMessage = new HydraMessage { Source = ThisParty, Destination = RemoteParty, Topic = Topic, Attachments = attachments,
                                                  Handle = Handle, Seq = LastSendSeq + 1, Data = _serializer.Serialize(message) };
            var res = _hydraService.Send(hydraMessage);
            // Increment LastSendSeq after sending in case the Send fails.
            LastSendSeq++;
            return res;
        }

        /// <summary>
        /// Send an augmented message to the other end of the conversation
        /// </summary>
        /// <param name="message">The augmented message to send.</param>
        /// <returns>The id of the message sent.</returns>
        public IMessageId Send(AugmentedMessage<TMessage> message)
        {
            return Send(message.Message, message.Attachments);
        }

        #region Implementation of IObservable<out Notification<AugmentedMessage<TMessage>>>

        public IDisposable Subscribe(IObserver<Notification<AugmentedMessage<TMessage>>> observer)
        {
            return _subject.Subscribe(observer);
        }

        #endregion

        #region Implementation of IDisposable

        // See http://msdn.microsoft.com/en-us/library/ms244737.aspx

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (disposing) {
                // free managed resources
                _done = true;
                _subject.Dispose();
                if (DoneEvent != null) DoneEvent(this);
            }
            // free native resources if there are any.
        }

        #endregion

    }
}

File: Java/hydra/src/main/java/uk/co/shastra/hydra/conversations/Conversation.java
package uk.co.shastra.hydra.conversations;

import rx.Notification;
import rx.Observable;
import rx.subjects.PublishSubject;
import rx.subjects.Subject;
import uk.co.shastra.hydra.messaging.HydraMessage;
import uk.co.shastra.hydra.messaging.HydraService;
import uk.co.shastra.hydra.messaging.messageids.MessageId;
import uk.co.shastra.hydra.messaging.serializers.Serializer;
import uk.co.shastra.hydra.messaging.utils.EventHandlerNoData;
import uk.co.shastra.hydra.messaging.utils.EventNoData;

/**
 * A Conversation is an exchange of messages between one sender and one recipient. Clients can engage in multiple simultaneous conversations
 * and messages on each are separate.
 *  
 * @author Nick North
 *
 * @param <TMessage> Type of messages comprising the Conversation.
 */
public class Conversation<TMessage> {

	private Serializer<TMessage> serializer;
	private HydraService hydraService;
    private Subject<Notification<TMessage>, Notification<TMessage>> subject = PublishSubject.create();
    private boolean done = false;
    
	private String thisParty;
	private String remoteParty;
	private String topic;
	private String handle;
	private long lastSendSeq;
	private long lastRecvSeq;
	private boolean checkSeq;
	
	private EventNoData doneEvent = new EventNoData();
    /**
     * Add a handler for the Done event, raised when a Conversation ends.
     * 
     * @param handler Object invoked when the Conversation ends.
     */
    public void addDoneHandler(EventHandlerNoData handler) { doneEvent.addHandler(handler); }
    /**
     * Remove a handler for the Done event.
     * 
     * @param handler Handler to be removed.
     */
    public void removeDoneHandler(EventHandlerNoData handler) { doneEvent.removeHandler(handler); }

	/**
	 * @return Identifier of this end of the Conversation.
	 */
	public String getThisParty() { return thisParty; }
	/**
	 * @return Identifier of the other end of the Conversation.
	 */
	public String getRemoteParty() { return remoteParty; }
	/**
	 * @return The Conversation topic. Defaults to the canonical name of the Conversation message type.
	 */
	public String getTopic() { return topic; }
	/**
	 * @return The Conversation handle. A GUID.
	 */
	public String getHandle() { return handle; }
	/**
	 * @return The sequence number of the last message sent. The first message has sequence number 1.
	 */
	public long getLastSendSeq() { return lastSendSeq; }
	/**
	 * @return The sequence number of the last message received.
	 */
	public long getLastRecvSeq() { return lastRecvSeq; }
	/**
	 * @return Whether out-of-sequence messages cause an error.
	 */
	public boolean isCheckSeq() { return checkSeq; }
	/**
	 * @param checkSeq Whether to check message sequence numbers.
	 */
	public void setCheckSeq(boolean checkSeq) { this.checkSeq = checkSeq; }
	
	void baseInit(HydraService hydraService, String thisParty, String remoteParty, String topic, String handle, Serializer<TMessage> serializer)
    {
        this.hydraService = hydraService;
        this.thisParty = thisParty;
        this.remoteParty = remoteParty;
        this.topic = topic;
        this.handle = handle;
        this.serializer = serializer;
    }
    
    void onNext(long seq, Notification<TMessage> message)
    {
        if (checkSeq && seq != lastRecvSeq + 1) {
            // Tell the client, but carry on and process the message.
        	subject.onError(new Exception(String.format("Sequence error. Expected %1$s, received %2$s", lastRecvSeq + 1, seq)));
        }
        lastRecvSeq = seq;
        subject.onNext(message);
    }
    
    /**
     * Send a message to the other end of the conversation
     * 
     * @param message The message to send
     * @return The id of the message sent
     * @throws Exception 
     */
    public MessageId send(TMessage message) throws Exception
    {
        if (done) return null;

        HydraMessage hydraMessage = new HydraMessage();
        hydraMessage.setSource(thisParty);
        hydraMessage.setDestination(remoteParty);
        hydraMessage.setTopic(topic);
        hydraMessage.setHandle(handle);
        hydraMessage.setSeq(lastSendSeq + 1);
        hydraMessage.setData(serializer.serialize(message));
        
        MessageId res = hydraService.send(hydraMessage);
        // Increment LastSendSeq after sending in case the Send fails.
        lastSendSeq++;
        return res;
    }
    
    /**
     * @return The Observable giving messages in this Conversation.
     */
    public Observable<Notification<TMessage>> getObservable() { return subject; }
    
    /**
     * Dispose of resources used by this Conversation.
     */
    public void close()
    {
        done = true;
        doneEvent.raise(this);
    }
}
